lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,217 @@
+            connection_name = (
+                "default"
+                if self._connection_name == "snowflake"
+                else self._connection_name
+            )
+            return snowflake.connector.connect(
+                connection_name=connection_name,
+                **kwargs,
+            )
+
+        # If we have a version of the Snowflake Python Connector installed that predates
+        # the ConfigManager, look for credentials in  ~/.snowsql/config as a last
+        # resort.
+        snowsql_config = load_from_snowsql_config_file(self._connection_name)
+        conn_kwargs = {**snowsql_config, **kwargs}
+
+        for p in _REQUIRED_CONNECTION_PARAMS:
+            if p not in conn_kwargs:
+                raise StreamlitAPIException(f"Missing Snowflake connection param: {p}")
+
+        return snowflake.connector.connect(**conn_kwargs)
+
+    def query(
+        self,
+        sql: str,
+        *,  # keyword-only arguments:
+        ttl: float | int | timedelta | None = None,
+        show_spinner: bool | str = "Running `snowflake.query(...)`.",
+        params=None,
+        **kwargs,
+    ) -> pd.DataFrame:
+        """Run a read-only SQL query.
+
+        TODO(vdonato): Finish this docstring.
+        """
+        # TODO(vdonato): Make our error handling more specific if possible. This may be
+        # difficult to do given the limited documentation on the different connector
+        # error subclasses + how many there are.
+        from snowflake.connector import Error as SnowflakeError
+        from tenacity import (
+            retry,
+            retry_if_exception_type,
+            stop_after_attempt,
+            wait_fixed,
+        )
+
+        @retry(
+            after=lambda _: self.reset(),
+            stop=stop_after_attempt(3),
+            reraise=True,
+            retry=retry_if_exception_type(SnowflakeError),
+            wait=wait_fixed(1),
+        )
+        @cache_data(
+            show_spinner=show_spinner,
+            ttl=ttl,
+        )
+        def _query(sql: str) -> pd.DataFrame:
+            cur = self._instance.cursor()
+            cur.execute(sql)
+            return cur.fetch_pandas_all()
+
+        return _query(sql)
+
+    def write_pandas(
+        self,
+        df: pd.DataFrame,
+        table_name: str,
+        database: str | None = None,
+        schema: str | None = None,
+        chunk_size: int | None = None,
+        **kwargs,
+    ) -> tuple[  # This return type is somewhat horrifying but is the return type of pandas_tools.write_pandas, so we keep it consistent here.
+        bool,
+        int,
+        int,
+        Sequence[
+            tuple[
+                str,
+                str,
+                int,
+                int,
+                int,
+                int,
+                str | None,
+                int | None,
+                int | None,
+                str | None,
+            ]
+        ],
+    ]:
```

Comment:
Note to self: Take a look at whether we want to just return a subset of these values for the Streamlit version

Reply:
@vdonato here's the current return value for a simple insert of 3 rows:

```python
import streamlit as st
import pandas as pd

conn = st.connection("snowflake")
conn.cursor().execute("CREATE TABLE IF NOT EXISTS MY_TABLE_FOO(IDENT NUMBER, AMOUNT NUMBER)")
df = pd.DataFrame([[1, 2],[3, 4], [5, 6]], columns=["IDENT", "AMOUNT"])
res = conn.write_pandas(df, "MY_TABLE_FOO")
st.write(res)
```
`(True, 1, 3, [('xxxxx/file0.txt', 'LOADED', 3, 3, 1, 0, None, None, None, None)])`

In Snowflake example docs for the function, they even drop that last sequence/tuple:
`success, nchunks, nrows, _ = write_pandas(cnx, df, 'customers')`

Why don't we follow the same pattern and just return a 3 element tuple `success, nchunks, nrows,` (I would almost argue that we reduce it down further to just return `nrows` and do `None` if `success==False`, but maybe that's too simplified)

Reply:
Yeah, I'm in favor of just retaining the first three elements of the tuple for this one